### PR TITLE
Updated CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by ascent12**
Changed the contribution guidelines to be more consistent with the style that wlroots has evolved into, partially mentioned in #93.
This is my current interpretation of what the style is actually supposed to be. Correct me if I'm wrong.

I've removed some of the things which was duplicating information from the kernel style, and am mainly highlighting the differences from it. We just defer to the kernel style in cases not mentioned here.

I removed the code example, as it had references to very sway specific things, and it didn't actually follow the style being talked about. Not to mention that I think the code was probably a bit too complicated for an example.  
I looked around the wlroots codebase for a replacement, but I couldn't find a reasonably self-contained example that showcased the style and not have some other problem with it.
We can add the example back when the code gets cleaned up a little bit.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -2,203 +2,138 @@
...
+they should be written as a pair of one of two forms:
+* `init`/`finish`: These initialize/deinitialize a type, but are **NOT**
+responsible for allocating it. They should accept a pointer to some
+pre-allocated memory (e.g. a member of a struct).
+* `create`/`destroy`: These also initialize/deinitialize, but will return a
+pointer to a `malloc`ed chunk of memory, and will `free` it in `destroy`.
+
+A destruction function should always be able to accept a NULL pointer or an
+otherwise uninitialised value and exit cleanly; this simplifies error handling
+a lot.
```
Comment (martinetd) [inline CONTRIBUTING.md]: uninitialized is a bit rough. Can't really handle garbage. zeroed?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -2,203 +2,138 @@
...
-	registry_teardown(registry);
-	return 0;
+For long lines that need to be broken, the continuation line should be indented
+with an additional tab.  
+If the line being broken is opening a new block (functions, if, while, etc.),
+the continuation line should be indented with two tabs, so they can't be
+misread as being part of the block.
+```c
+really_long_function(argument1, argument2, ...,
+	argument3, argument4);
```
Comment (martinetd) [inline CONTRIBUTING.md]: hm I was using two tabs for that as well - that's what we have in most headers.
Reasoning (for me) is that it allows a straight copy from definition to declaration.

Comment (ascent12) [inline CONTRIBUTING.md]: I've seen both in use, so it'll be good if we can decide on a style and have everybody consistently use it.

File: CONTRIBUTING.md (hunk 1 above)
Comment (ascent12) [inline CONTRIBUTING.md]: Yeah, I should probably mention the zeroing/calloc somewhere.

File: CONTRIBUTING.md (hunk 2 above)
Comment (SirCmpwn) [inline CONTRIBUTING.md]: Definition -  two tabs

Usage - one tab

Comment (ascent12) [inline CONTRIBUTING.md]: That seems like an awfully specific rule. I would prefer two tabs across the board.

Comment (SirCmpwn) [inline CONTRIBUTING.md]: There's a reason for it. Definitions indent an additional time to make the indent distinct from the body:

```c
void foobar(...,
        ...) {
    ...
}
```

Whereas calls use only one indent so they can fit more arguments on one line.

Comment (ascent12) [inline CONTRIBUTING.md]: Oh, so you just mean function definitions?
I thought you meant all definitions, including normal variables.
```c
int really_long_variable_name = really_long_initialiser(...,
        ...);
really_long_function_usage(...,
    ...);
```
Which I thought was kind of pointless.

If it's just functions, then that's covered by the "If the line being broken is opening a new block".

Comment (SirCmpwn) [inline CONTRIBUTING.md]: Yeah just function defs. Thanks for clarifying.

**Comment by SirCmpwn** [comment]
bump

**Comment by ascent12** [comment]
Updated.

**Comment by SirCmpwn** [comment]
Thanks!

**Pull request merged**